Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+extension HTTPConnectionPool {
+    struct HTTP1ConnectionState {
+        enum State {
+            case starting(Waiter?)
+            case active(Connection, isAvailable: Bool, lastReturn: NIODeadline)
+            case failed
+            case closed
+        }
+
+        private(set) var state: State
+        let eventLoop: EventLoop
+        let connectionID: Connection.ID
+
+        init(connectionID: Connection.ID, eventLoop: EventLoop, waiter: Waiter) {
+            self.connectionID = connectionID
+            self.eventLoop = eventLoop
+            self.state = .starting(waiter)
+        }
+
+        var isStarting: Bool {
+            switch self.state {
+            case .starting:
+                return true
+            case .failed, .closed, .active:
+                return false
+            }
+        }
+
+        var isAvailable: Bool {
+            switch self.state {
+            case .active(_, let isAvailable, _):
+                return isAvailable
+            case .starting, .failed, .closed:
+                return false
+            }
+        }
+
+        var isLeased: Bool {
+            switch self.state {
+            case .active(_, let isAvailable, _):
+                return !isAvailable
+            case .starting, .failed, .closed:
+                return false
+            }
+        }
+
+        var availableAndLastReturn: NIODeadline? {
+            switch self.state {
+            case .active(_, true, let lastReturn):
+                return lastReturn
+            case .active(_, false, _):
+                return nil
+            case .starting, .failed, .closed:
+                return nil
+            }
+        }
+
+        mutating func started(_ connection: Connection) -> Waiter? {
+            guard case .starting(let waiter) = self.state else {
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+            self.state = .active(connection, isAvailable: true, lastReturn: .now())
+            return waiter
+        }
+
+        mutating func failed() -> Waiter? {
+            guard case .starting(let waiter) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .failed
+            return waiter
+        }
+
+        @discardableResult
+        mutating func lease() -> Connection {
+            guard case .active(let conn, isAvailable: true, let lastReturn) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .active(conn, isAvailable: false, lastReturn: lastReturn)
+            return conn
+        }
+
+        mutating func release() {
```

Comment:
because further down we pick the connection based on its id:

```swift
        mutating func http1ConnectionReleased(_ connectionID: Connection.ID) -> Action {
            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
                preconditionFailure("A connection that we don't know was released? Something is very wrong...")
            }

            var connectionState = self.connections[index]
            connectionState.release()
```